### ISSUE TITLE
TypeVar incorrectly inferred from default value in `T | SentinelType = sentinel` pattern when used with stubs

### ISSUE THREAD

**Issue by mgrinshpon** [microsoft/pyright#11142]
**Describe the bug**

When a generic function has a parameter typed as `T | SentinelType = sentinel` where:
- `T` is a TypeVar constrained by other parameters
- `SentinelType` is a `Literal["..."]` type
- `sentinel` is the actual string literal value

Pyright incorrectly includes `str` in `T`'s inference, even when the caller does not provide that argument.

**Code or Screenshots**

- Pyright version: 1.1.406
- Python version: 3.12
- Steps to reproduce
    1. Make these files in a directory
    2. Run uvx pyright
- File structure
    - mylib.py
    - mylib.pyi
    - test_types.py
    - pyrightconfig.json

_mylib.py_

```python
no_default = "__no__default__"
def accumulate(binop, seq, initial=no_default):
    pass
```

_mylib.pyi_
```python
import collections.abc
import typing

no_default = "__no__default__"
type _NoDefaultType = typing.Literal["__no__default__"]

def accumulate[T](
    binop: typing.Callable[[T, T], T],
    seq: collections.abc.Iterable[T],
    initial: T | _NoDefaultType = no_default,
) -> collections.abc.Iterator[T]: ...
```

_test_types.py_
```python
from typing import assert_type
from mylib import accumulate

add = int.__add__
nums = [1, 2, 3]

# BUG: T should be inferred as `int` from binop and seq,
# but pyright incorrectly infers T as `int | str` 
# because the default value `no_default` is a string.
result = list(accumulate(add, nums))
_ = assert_type(result, list[int])
```

_pyrightconfig.json_
```json
{
  "pythonVersion": "3.12"
}
```

**VS Code extension or command-line**

Command line

**Workaround**

I can use overloads to separate the issue but this does make it challenging for stub libraries. Workaround:

```python
@typing.overload
def accumulate[T](
    binop: Callable[[T, T], T],
    seq: Iterable[T],
) -> Iterator[T]: ...

@typing.overload
def accumulate[T](
    binop: Callable[[T, T], T],
    seq: Iterable[T],
    initial: T,
) -> Iterator[T]: ...

def accumulate[T](
    binop: Callable[[T, T], T],
    seq: Iterable[T],
    initial: T | _NoDefaultType = no_default,
) -> Iterator[T]: ...
```

**Comment by erictraut** [comment]
I'm not able to repro the problem as reported. You've included a bunch of extraneous details, so here's an attempt to reduce the problem to an MRE:

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQCSM1IlN9TLRMu9AMkmasSAKDEATasFjUAzjADaAFQC6AChFQtUIgC5CpclWoLl9VeZX0K%2Bzq2PKrUAMb6lUAD5QBQ0goBEAB7%2BKlAAvFBB-iIAlFAAtAB8jL48II76AHTZYiJE4uLhmCgwmQD6ZfniFSIoAK4QskUKAIz0AEz0AMwqYiBydST4EczyalX09Y0xIv0AbtSkZfAI1Gr9soMwcVAAxClcaQqoML1AA)

```python
from typing import Callable, Iterable, Iterator, Literal

def test[T](
    a: Callable[[T, T], T], b: Iterable[T], c: T | Literal["x"] = "x"
) -> Iterator[T]: ...

add = int.__add__
nums = [1, 2, 3]

result = test(add, nums)
reveal_type(result)  # Iterator[int]
```

Pyright is (I think correctly) evaluating `result` as `Iterator[int]` in this case. Are you seeing a different result?

**Comment by mgrinshpon** [comment]
Hey @erictraut , I think the issue relates specifically to how untyped `.py` file interacts with the generic in the `.pyi` file and type narrowing. The `.pyi` file says "this is `T | Literal["x"]` but the `.py` file says "this is `T | str` and that means your `assert_type` is wrong."

Would it be helpful if I reproduced the issue in a repo and demonstrated the failure in GHA?

**Comment by erictraut** [comment]
It would be helpful if you could provide a minimal reproducible example (MRE) — preferably in the pyright playground if possible.

**Comment by mgrinshpon** [comment]
Gotcha and done. Since the problem only occurs when there are multiple files involved (the .py file, the .pyi file, and a file calling the .py file looking at the .pyi file for type hinting), I could not use the pyright playground.

* Repo: https://github.com/mgrinshpon/pyright-11142-mre
* CI failure demonstrating the problem observed with pyright: https://github.com/mgrinshpon/pyright-11142-mre/actions/runs/19776512068/job/56669908034
* [mylib.py](https://github.com/mgrinshpon/pyright-11142-mre/blob/main/mylib.py)
* [mylib.pyi](https://github.com/mgrinshpon/pyright-11142-mre/blob/main/mylib.pyi)
* [the test file](https://github.com/mgrinshpon/pyright-11142-mre/blob/main/test_types.py)

**Comment by erictraut** [comment]
Thanks for the MRE.

The problem is that the symbol `no_default` in your code is a variable with no declared type. Its type is therefore inferred as `str`. I think your intent is for `no_default` to be a constant and for its type to be `Literal["__no__default__"]`.

You can fix this by either:
1. Declaring the type of `no_default`: `no_default: Literal["__no__default__"] = "__no__default__"`.
2. Using the `Final` qualifier to indicate that `no_default` is constant: `no_default: Final = "__no__default__"`
3. Rename `no_default` to all caps, in which case pyright will treat it as a constant and use Final-like inference rules: `NO_DEFAULT = "__no__default__"`

Option 2 is probably the cleanest. Option 1 and 2 should both work for all type checkers. Option 3 may not work with all type checkers, so it's probably the least appealing choice.

In any case, this isn't a bug in pyright, so I'm going to close the issue.
